vm/src/builtins/range.rs

```diff
@@ -634,10 +634,14 @@ impl TryFromObject for RangeIndex {
         match_class!(match obj {
             i @ PyInt => Ok(RangeIndex::Int(i)),
             s @ PySlice => Ok(RangeIndex::Slice(s)),
-            obj => Err(vm.new_type_error(format!(
-                "sequence indices be integers or slices, not '{}'",
-                obj.class().name,
-            ))),
+            obj => match vm.to_index(&obj) {
+                Ok(val) => Ok(RangeIndex::Int(val)),
+                Err(_) => Err(vm.new_type_error(format!(
+                    "sequence indices be integers or slices or classes that override __index__ operator, not '{}'",
+                    obj.class().name
+                ))),
+            },
```

Comment:
```suggestion
            obj => {
                let val = vm.to_index(&obj).map_err(|_| vm.new_type_error(format!(
                    "sequence indices be integers or slices or classes that override __index__ operator, not '{}'",
                    obj.class().name
                )));
                RangeIndex::Int(val)
            }
```

Your version also looks very good, but we also conventionally write it like this to show the success flow better.

Reply:
Oh, you are right. it is better seems great. But in my environment, rust complained about type mismatching like below.

>    Compiling rustpython-vm v0.1.2 (/home/snowapril/RustPython/vm)
> error[E0308]: mismatched types
>    --> vm/src/builtins/range.rs:642:33
>     |
> 642 |                 RangeIndex::Int(val)
>     |                                 ^^^ expected struct `pyobjectrc::PyRef`, found enum `Result`
>     |
>     = note: expected struct `pyobjectrc::PyRef<_>`
>                  found enum `Result<pyobjectrc::PyRef<_>, pyobjectrc::PyRef<PyBaseException>>`
> 
> error[E0308]: mismatched types
>    --> vm/src/builtins/range.rs:642:17
>     |
> 633 |     fn try_from_object(vm: &VirtualMachine, obj: PyObjectRef) -> PyResult<Self> {
>     |                                                                  -------------- expected `Result<RangeIndex, pyobjectrc::PyRef<PyBaseException>>` because of return type
> ...
> 642 |                 RangeIndex::Int(val)
>     |                 ^^^^^^^^^^^^^^^^^^^^
>     |                 |
>     |                 expected enum `Result`, found enum `RangeIndex`
>     |                 help: try using a variant of the expected enum: `serde::__private::Ok(RangeIndex::Int(val))`
>     |
>     = note: expected enum `Result<RangeIndex, pyobjectrc::PyRef<PyBaseException>>`
>                found enum `RangeIndex`
> 
> error: aborting due to 2 previous errors
> 
> For more information about this error, try `rustc --explain E0308`.
> error: could not compile `rustpython-vm`
> 
> To learn more, run the command again with --verbose.

I pushed the second version of the code below. could you review it one more time? thanks.